src/main/java/com/example/java_tutorial/MainController.java

```diff
@@ -1,59 +1,114 @@
 import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
 @RequestMapping(path = "")
 public class MainController {
 
-  // ② Create a logger for this class
+  private final NorthwindOlingoService service;
+
   private static final Logger logger = LoggerFactory.getLogger(MainController.class);
 
-  private static final String DESTINATION_NAME = "SecondAppSub";
-  private static final String REL_URL = "/hello";
+  private static final String NORTHWIND_DESTINATION = "Northwind";
+  private static final String SECOND_APP_DESTINATION = "SecondAppSub";
+  private static final String SECOND_APP_URL = "/hello";
+
+  private final DefaultNorthwindService northwindService =
+      new DefaultNorthwindService()
+          .withServicePath("/V4/Northwind/Northwind.svc/");
+
+  public MainController(NorthwindOlingoService service) {
+    this.service = service;
+  }
 
   @GetMapping(path = "")
   public ResponseEntity<String> readAll(@AuthenticationPrincipal Token token) {
-    // ③ Log entry into the method
-    logger.info("readAll() called; user authorities: {}", token.getAuthorities());
+    try {
+      if (!token.getAuthorities().contains(new SimpleGrantedAuthority("Display"))) {
+        logger.warn("Access denied – missing 'Display' scope for token {}", token.getAppToken());
+        return ResponseEntity
+            .status(HttpStatus.FORBIDDEN)
+            .body("Access denied: missing 'Display' scope");
+      }
 
-    // ④ Check scopes and log a warning on failure
-    if (!token.getAuthorities().contains(new SimpleGrantedAuthority("Display"))) {
-      logger.warn("Access denied – missing 'Display' scope for token: {}", token.getAppToken());
-      throw new NotAuthorizedException("This operation requires \"Display\" scope");
+      logger.debug("User authorized");
+      return ResponseEntity.ok("Hello World!");
+    } catch (Exception e) {
+      logger.error("Unexpected error in readAll()", e);
+      return ResponseEntity
+          .status(HttpStatus.INTERNAL_SERVER_ERROR)
+          .body("An unexpected error occurred.");
     }
-
-    // ⑤ Optionally log before returning
-    logger.debug("Returning Hello World response");
-    return new ResponseEntity<>("Hello World!", HttpStatus.OK);
   }
 
   @GetMapping("/call-second")
   public String callSecondApp() throws IOException {
-    HttpDestination destination = DestinationAccessor.getLoader().tryGetDestination(DESTINATION_NAME)
-        .get().asHttp();
-    HttpClient client = HttpClientAccessor.getHttpClient(destination);
-    HttpGet httpGet = new HttpGet(REL_URL);
-    HttpResponse httpResponse = client.execute(httpGet);
-    return IOUtils.toString(httpResponse.getEntity().getContent(), StandardCharsets.UTF_8);
+    HttpDestination dest = DestinationAccessor
+        .getLoader()
+        .tryGetDestination(SECOND_APP_DESTINATION)
```

Comment:
This also should not be in controller. Please leave all the logic inside service level, only Response body consturction and another services call's should be in controller